### PR TITLE
SCC-HOIST: Hoist variables as `kwargs` (optionally)

### PR DESCRIPTION
**Pull request by MichaelSt98**
Hoist variables as `kwargs` (optionally)

### CODE REVIEW

**Review by reuterbal — CHANGES_REQUESTED** [review]
Thanks, that looks very useful! 
Just a little remark about parameterising tests, and a question on whether we may run into problems of `None + tuple()` with `call.kwarguments`.

File: tests/test_transform_hoist_variables.py [hunk 1]
```diff
@@ -104,22 +107,26 @@ def check_arguments(scheduler, subroutine_arguments, call_arguments, include_dev
...
                 assert call.arguments == call_arguments["device2"]
+                assert call.kwarguments == call_kwarguments["device2"]
         # device 2
         item = scheduler['subroutines_mod#device2']
         assert [arg.name for arg in item.ir.arguments] == subroutine_arguments["device2"]
 
 
 @pytest.mark.parametrize('frontend', available_frontends())
-def test_hoist(here, frontend, config):
+@pytest.mark.parametrize('as_kwarguments', ['False', 'True'])
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply tests/test_transform_hoist_variables.py:128]: There is no reason ... 😄 Thanks

File: loki/transform/transform_hoist_variables.py [hunk 2]
```diff
@@ -307,8 +315,13 @@ def driver_call_argument_remapping(self, routine, call, variables):
             Call object to which hoisted variables will be added.
         variables : tuple of :any:`Variable`
             The tuple of variables to be declared.
+        as_kwarguments : boolean
+            Whether to pass the hoisted arguments as `args` or `kwargs`.
         """
         # pylint: disable=unused-argument
+        if self.as_kwarguments:
+            new_kwargs = tuple((a.name, v.clone(dimensions=None)) for (a, v) in variables)
+            return call.clone(kwarguments=call.kwarguments + new_kwargs)
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply loki/transform/transform_hoist_variables.py:324]: Good point, seems like it's possible that `call.kwarguments` could be `None`

**Review by reuterbal — APPROVED** [review]

**Review by mlange05 — APPROVED** [review]
Looks good. Many thanks for your patience. GTG :shipit:

**Pull request merged**